exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -218,4 +302,141 @@ private void registerOperatorsWithoutInference(DrillOperatorTable operatorTable)
       }
     }
   }
+
+  /**
+   * Function registry holder. Stores function implementations by jar name, function name.
+   */
+  private class GenericRegistryHolder<T, U> {
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final AutoCloseableLock readLock = new AutoCloseableLock(readWriteLock.readLock());
+    private final AutoCloseableLock writeLock = new AutoCloseableLock(readWriteLock.writeLock());
+
```

Comment:
I use ReentrantReadWriteLock to perform locking, and then wrap its read / write locks into AutoCloseableLock. It should work as documented [1]: allow multiple reads at the same time, restrict to one write at a time, disallow reads and writes at the same time.

[1] https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html